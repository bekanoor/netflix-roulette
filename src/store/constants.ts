export const VIEW_PAGE_HEAD_ERROR = 'VIEW_PAGE_HEAD_ERROR'
export const VIEW_PAGE_MAIN_ERROR = 'VIEW_PAGE_MAIN_ERROR'
export const MAIN_PAGE_ERROR = 'MAIN_PAGE_ERROR'