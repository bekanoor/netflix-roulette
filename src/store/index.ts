export * from './actions'
export *  from './store'
export * from './moviesReducer'
export * from './searchParamReducer'
