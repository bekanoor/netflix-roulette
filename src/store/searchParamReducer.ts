import { actionType } from '../models'

type defaultStateType = {
  searchBy: string
  filterBy: string
  searchQuery: string
  isButton: boolean
}

const defaultState: defaultStateType = {
  searchBy: 'title',
  filterBy: 'vote_average',
  searchQuery: '',
  isButton: false,
}

const searchParamReducer = (state = defaultState, action: actionType) => {
  switch (action.type) {
    case 'SET_SEARCH_BY':
      return { ...state, searchBy: action.payload }
    case 'SET_FILTER_BY':
      return { ...state, filterBy: action.payload }
    case 'SET_SEARCH_QUERY':
      return { ...state, searchQuery: action.payload }
    case 'SET_BUTTON_STATE':
      return { ...state, isButton: action.payload }
    default:
      return state
  }
}

export { searchParamReducer }
