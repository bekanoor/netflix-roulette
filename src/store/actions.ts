import { Movie } from 'src/models'

export const setSearchType = (value: 'title' | 'genres') => {
  return { type: 'SET_SEARCH_BY', payload: value }
}

export const setFilterType = (value: string) => {
  return { type: 'SET_FILTER_BY', payload: value }
}

export const setSearchQuery = (value: string) => {
  return { type: 'SET_SEARCH_QUERY', payload: value }
}

export const setSearchButton = (value: boolean) => {
  return { type: 'SET_BUTTON_STATE', payload: value }
}

export const setData = (data: any) => {
  return { type: 'SET_DATA', payload: data }
}

export const setLoading = (status: boolean) => {
  return { type: 'SET_LOADING', payload: status }
}

export const setSelectedMovie = (value: Array<Movie>) => {
  return { type: 'SET_SELECTED_MOVIE_ID', payload: value }
}

export const setSameGenreMovies = (value: Array<Movie>) => {
  return { type: 'SET_SAME_GENRE_MOVIES', payload: value }
}
