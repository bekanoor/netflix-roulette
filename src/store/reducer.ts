import { actionType, stateTypes } from 'src/models'

const defaultState: stateTypes = {
  searchType: 'title',
  filterType: 'rating',
  searchInput: '',
  searchButton: 'disable',
  data: {
    data: [],
  },
  movieID: []
}

const reducer = (state: stateTypes = defaultState, action: actionType) => {
  switch (action.type) {
    case 'SET_SEARCH_TYPE':
      return { ...state, searchType: action.payload }
    case 'SET_FILTER_TYPE':
      return { ...state, filterType: action.payload }
    case 'SET_SEARCH_INPUT':
      return { ...state, searchInput: action.payload }
    case 'SET_SEARCH_BUTTON':
      return { ...state, searchButton: action.payload }
    case 'GET_DATA':
      return { ...state, data: action.payload }
    case 'SET_MOVIE_ID':
      return { ...state, movieID: [...state.movieID, action.payload] }
    case 'GET_PREV_STATE':
      return { ...state, movieID: state.movieID.slice(0, -1) }
    default:
      return state
  }
}

export { reducer }
