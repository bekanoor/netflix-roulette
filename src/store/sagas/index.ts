import {
  fetchDataByGenre,
  fetchDataById,
  fetchMovies,
  fetchMoviesBySearchType,
} from '../../utils/'
import {
  finishSetMovies,
  finishSetSameGenreMovies,
  setMainPageError,
  setSelectedMovie,
  setViewPageHeadError,
} from '../actions/actions'
import {
  SET_NEW_TAB_MOVIE,
  START_SET_MOVIES,
  START_SET_SAME_GENRE_MOVIES,
} from '../constants'

import { call, put, takeEvery } from 'redux-saga/effects'
import { Movie } from 'src/models'
import { useAppSelector } from 'src/hook'

type HandleMoviesType = {
  type: string
  payload: ['query', 'searchBy', 'filterBy']
}

export function* handleMovies(action: HandleMoviesType) {
  if (action?.payload === undefined) {
    const data: Array<Movie> = yield call(fetchMovies)

    yield put(finishSetMovies(data))
  } else {
    const [query, searchBy, filterBy] = action.payload
    const data: Array<Movie> = yield call(
      fetchMoviesBySearchType,
      query,
      searchBy,
      filterBy
    )

    if (!data.length) {
      yield put(setMainPageError())
      yield put(finishSetMovies([]))
      return
    }

    yield put(finishSetMovies(data))
  }
}

type HandleTransitionFromMainPage = {
  type: string
  payload: string
}

export function* handleTransitionFromMainPage(
  action: HandleTransitionFromMainPage
) {
  const data: Array<Movie> = yield call(fetchDataByGenre, action.payload)

  yield put(finishSetSameGenreMovies(data))
}

type ActionType = {
  type: string
  payload: [string, Array<Movie>]
}

export function* handleViewPageHeader(action: ActionType) {
  try {
    const [id, movies] = action.payload

    const [cachedMovie] = [
      movies.find((item) => item.id === +id),
    ] as Array<Movie>

    if (cachedMovie) {
      yield put(setSelectedMovie([cachedMovie]))
      return
    }

    const movie: Movie = yield call(fetchDataById, id)
    //here
    yield put(setSelectedMovie([movie]))
  } catch (err) {
    yield put(setViewPageHeadError())
  }
}

export function* handleViewPageMain(action: ActionType) {
  try {
    const [id] = action.payload
    const movie: Movie = yield call(fetchDataById, id)
    const { genres } = movie
    const [genre] = genres

    const sameGenreMovies: Array<Movie> = yield call(fetchDataByGenre, genre)

    yield put(finishSetSameGenreMovies(sameGenreMovies))
  } catch (err) {
    yield put(setViewPageHeadError())
  }
}

export function* watchSaga() {
  yield takeEvery(START_SET_MOVIES, handleMovies)
  yield takeEvery(START_SET_SAME_GENRE_MOVIES, handleTransitionFromMainPage)
  yield takeEvery(SET_NEW_TAB_MOVIE, handleViewPageHeader)
  yield takeEvery(SET_NEW_TAB_MOVIE, handleViewPageMain)
}

export function* rootSaga() {
  yield watchSaga()
}
