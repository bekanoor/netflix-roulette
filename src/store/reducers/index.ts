import { errorReducer } from './errorsReducer'
import { moviesReducer } from './moviesReducer'
import { searchParamReducer } from './searchParamReducer'

export { errorReducer, moviesReducer, searchParamReducer }
