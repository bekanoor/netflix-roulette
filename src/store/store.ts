import { reducer } from './reducer'

import { createStore } from 'redux'


const store = createStore(reducer)

export { store }

export type RootState = ReturnType<typeof store.getState >
export type AppDispatch = typeof store.dispatch  
