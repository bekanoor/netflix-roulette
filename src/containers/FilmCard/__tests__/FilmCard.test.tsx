import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { FilmCard } from '../'
import { Provider } from 'react-redux'
import { store } from '../../../store'
import { BrowserRouter } from 'react-router-dom'

describe('test film card', () => {
  test('should render film card correctly', () => {
    const { container } = render(
      <BrowserRouter>
        <Provider store={store}>
          <FilmCard cover='' filmTitle='' releaseDate='2018' genre='' id={0} />
        </Provider>
      </BrowserRouter>
    )

    expect(container).toMatchSnapshot()
  })

  test('dom elements really exist', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <FilmCard cover='' filmTitle='' releaseDate='2018' genre='' id={0} />
        </Provider>
      </BrowserRouter>
    )

    const card = screen.getByTestId('film-card-test')
    expect(card).toBeInTheDocument
  })
})
