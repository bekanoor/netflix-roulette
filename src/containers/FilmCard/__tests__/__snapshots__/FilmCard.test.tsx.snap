// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test film card should render film card correctly 1`] = `
<div>
  <a
    href="/view-page/0-"
  >
    <div
      class="film-card"
      data-testid="film-card-test"
      id="0"
    >
      <div
        class="film-card__image-container"
      >
        <img
          alt="image"
          class="film-card__image"
          src=""
        />
      </div>
      <div
        class="film-card__title-container"
      >
        <p
          class="film-card__title"
        />
        <button
          class="film-card__release-date"
        >
          2018
        </button>
      </div>
      <p
        class="film-card__genre"
      />
    </div>
  </a>
</div>
`;
